fenzo-core/src/main/java/com/netflix/fenzo/BaseShortfallEvaluator.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2017 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.fenzo;
+
+import com.netflix.fenzo.functions.Func1;
+import com.netflix.fenzo.queues.QueuableTask;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+abstract class BaseShortfallEvaluator implements ShortfallEvaluator {
+    private static final long TOO_OLD_THRESHOLD_MILLIS = 10 * 60000; // should this be configurable?
+
+    private volatile Func1<QueuableTask, List<String>> taskToClustersGetter = null;
+    private final Map<String, Long> requestedForTasksSet = new HashMap<>();
+
+    protected TaskSchedulingService schedulingService = null;
+
+    @Override
+    public void setTaskToClustersGetter(Func1<QueuableTask, List<String>> getter) {
+        this.taskToClustersGetter = getter;
+    }
+
+    @Override
+    public abstract Map<String, Integer> getShortfall(Set<String> attrKeys, Set<TaskRequest> failures, AutoScaleRules autoScaleRules);
+
+    @Override
+    public void setTaskSchedulingService(TaskSchedulingService schedulingService) {
+        this.schedulingService = schedulingService;
+    }
+
+    protected void reset() {
```

Comment:
Is here na idea to consider a task only once, and give an external cluster enough time to scale up so the task can be scheduled. TOO_OLD_THRESHOLD_MILLIS value is aligned now with scaling pause duration. If this is always the case, do we need this extra time-management here?